### PR TITLE
Kusama: Make the current inflation formula adjustable. 

### PR DESCRIPTION
**Pull request by kianenigma**
This PR does not alter the inflation of Kusama, but instead make the current parameters of the system fully configurable. 

## Parameters

The parameters are as follows: 

```rust
/// Minimum inflation rate used to calculate era payouts.
pub static MinInflation: Perquintill = Perquintill::from_rational(25u64, 1000);

/// Maximum inflation rate used to calculate era payouts.
pub static MaxInflation: Perquintill = Perquintill::from_percent(10);

/// Ideal stake ratio used to calculate era payouts.
pub static IdealStake: Perquintill = Perquintill::from_percent(75);

/// Falloff used to calculate era payouts.
pub static Falloff: Perquintill = Perquintill::from_percent(5);

/// Whether to use auction slots or not in the calculation of era payouts, then we subtract
/// `num_auctioned_slots.min(60) / 200` from `ideal_stake`.
///
/// That is, we assume up to 60 parachains that are leased can reduce the ideal stake by a
/// maximum of 30%.
///
/// With the move to agile-coretime, this parameter does not make much sense and should
/// generally be set to false.
pub static UseAuctionSlots: bool = true;
```

All of the above are exactly the current values in Kusama, and leave everything unchanged, leading to roughly `1kKSM` minted for validators, and around `90KSM` for treasury. 

All of the above can be changed via the `Root` track only. 

Given that it is hard to come to consensus, I highly advise this PR (and a similar PR to Polkadot) to NOT alter any parameter and leave that to the token holders. 

Just to provide one example, if we set `UseAuctionSlots = false` and leave everything else un-changed, the amount minted per era would be: 

* `807KSM` for staking
* `238KSM` for treasury

> This, in my opinion, is the most non-controversial change as the incorporation of auctions in the inflation rate is already meaningless with agile-coretime. 

## Consideration for UIs 

This PR is shipped with a new runtime API that is only added to the Kusama runtime: `Inflation_experimental_inflation_info`. I hope this API is used by the UIs to show the inflation parameters of the network, rather than re-creating the Rust logic in the front-end.

### CODE REVIEW

**Comment by kianenigma** [comment]
Some updates: 

1. Depends on https://github.com/polkadot-fellows/runtimes/pull/381
2. Would also preferably depend on sdk 1.15 to get the new `relay_inflation` helper fn. 
3. Based on the last fellowship call, for now we will only add this to Kusama https://www.youtube.com/live/oWppBMrgU7M?si=b4O-G_mIEBp1tnV4&t=2137

**Comment by kianenigma** [comment]
Once https://github.com/polkadot-fellows/runtimes/pull/381 is merged, I shall merge https://github.com/kianenigma/runtimes/tree/kiz-inflation-with-update here.

**Comment by kianenigma** [comment]
Well, I called this simulation more or less in the middle of an era, so the era duration is assumed to be half :)) So the code adds up.

File: relay/kusama/src/lib.rs
Comment (ggwpez) [inline relay/kusama/src/lib.rs:1984]: Why is it using balances here and not excluding the disabled issuance with `Nis::issuance().other`?

File: relay/kusama/src/lib.rs
Comment (ggwpez) [inline relay/kusama/src/lib.rs:3073]: Downloading the state lazily would be much smarter -  no need to declare upfront what keys are required.

**Review by ggwpez — APPROVED** [review]

File: relay/kusama/src/lib.rs
Comment (kianenigma) [reply relay/kusama/src/lib.rs:3073]: yeah, in this case it is fine as nis pallet is mostly empty.

remote-ext still doesn't have lazy download, right? or are you suggesting it because it now has it :o?

File: relay/kusama/src/lib.rs
Comment (kianenigma) [inline relay/kusama/src/lib.rs:1984]:
```suggestion
		let stake_able_issuance = Nis::issuance().other;
```

Comment (kianenigma) [reply relay/kusama/src/lib.rs:1984]: great catch!

File: relay/kusama/src/lib.rs
Comment (ggwpez) [reply relay/kusama/src/lib.rs:3073]: >or are you suggesting it because it now has it :o?

Nah i think it still has not. But could be nice 🙃

File: relay/kusama/Cargo.toml
Comment (ggwpez) [inline relay/kusama/Cargo.toml:118]: Zepter not happy here since all deps need to be declared in the workspace.

File: relay/kusama/src/lib.rs
Comment (gpestana) [inline relay/kusama/src/lib.rs:1984]:
```suggestion
		let stakable_issuance = Nis::issuance().other;
```

nit but for consistency with the rest of the code.

File: relay/kusama/src/lib.rs
Comment (gpestana) [inline relay/kusama/src/lib.rs:1992]:
```suggestion
				.filter(|i| *i >= LOWEST_PUBLIC_ID.into())
```

File: relay/kusama/src/lib.rs
Comment (gpestana) [inline relay/kusama/src/lib.rs:1987]: I'd probably refactor the body of this if statement to its own method. It is used in the `fn era_payout` as well and future changes maybe lead to inconsistencies between the two `inflation` values in the era payout and in the runtime API, if devs are not careful. Especially since there are some magic numbers (e.g. `Perquintil (auctioned_slots.min(60), 200u64)`) in it.

File: relay/kusama/src/lib.rs
Comment (gpestana) [inline relay/kusama/src/lib.rs:2000]: Due to this assumption, maybe we should name the runtime API as `exp_inflation_predicted_info`?

File: relay/kusama/src/lib.rs
Comment (gpestana) [inline relay/kusama/src/lib.rs:710]: ping to remember before merging.

**Review by gpestana — APPROVED** [review]
LGTM modulo a few comments and suggestions and the benchmarks for the pallet parameters.

File: relay/kusama/src/lib.rs
Comment (kianenigma) [reply relay/kusama/src/lib.rs:1987]: Let's discuss this once I merge this and move on to Polkadot, which should follow soon :)

File: relay/kusama/src/lib.rs
Comment (kianenigma) [inline relay/kusama/src/lib.rs:747]: As a final step, moving this to a shared crate as it will be needed for Polkadot as well.

**Comment by kianenigma** [comment]
bot fmt

**Review by franciscoaguirre — APPROVED** [review]

File: relay/common/src/lib.rs
Comment (acatangiu) [inline relay/common/src/lib.rs:115]: would be better if you referenced the crate name and (min) version required, instead of this PR, otherwise it becomes a non-trivial chore to figure out when this can be removed and no one will do it :)

(same above)

**Review by acatangiu — APPROVED** [review]
high-level looks good, did not double-check the actual staking initial values coded in

File: relay/common/src/lib.rs
Comment (joepetrowski) [inline relay/common/src/lib.rs:74]:
```suggestion
	/// Legacy auction proportion, which, if not `None`, is subtracted from `ideal_stake`.
```

File: relay/common/src/lib.rs
Comment (joepetrowski) [inline relay/common/src/lib.rs:79]:
```suggestion
/// A specialized function to compute the inflation of the staking system, tailored for Polkadot
/// Relay Chains, such as Polkadot, Kusama, and Westend.
```

File: relay/kusama/src/lib.rs
Comment (joepetrowski) [inline relay/kusama/src/lib.rs:661]:
```suggestion
		/// Whether to use auction slots or not in the calculation of era payouts. If true, then we
		/// subtract `num_auctioned_slots.min(60) / 200` from `ideal_stake`.
```

File: relay/kusama/src/lib.rs
Comment (joepetrowski) [inline relay/kusama/src/lib.rs:666]:
```suggestion
		/// With the move to Agile Coretime, this parameter does not make much sense and should
```

**Pull request merged**
